npm/networkPolicyController.go

```diff
@@ -0,0 +1,492 @@
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the network policy resource with this namespace/name
+	netPolObj, err := c.netPolLister.NetworkPolicies(namespace).Get(name)
+
+	// (TODO): Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("Network Policy %s is not found, may be it is deleted", key)
+			// netPolObj is not found, but should need to check the RawNpMap cache with cachedNetPolKey.
+			// cleanUpNetworkPolicy method will take care of the deletion of a cached network policy if the cached network policy exists with cachedNetPolKey in our RawNpMap cache.
+			cachedNetPolKey := util.GetNSNameWithPrefix(key)
+			err = c.cleanUpNetworkPolicy(cachedNetPolKey, SafeToCleanUpAzureNpmChain)
+			if err != nil {
+				return fmt.Errorf("[syncNetPol] Error: %v when network policy is not found\n", err)
+			}
+		}
+		return err
+	}
+
+	// If DeletionTimestamp of the netPolObj is set, start cleaning up lastly applied states.
+	// This is early cleaning up process from updateNetPol event
+	if netPolObj.ObjectMeta.DeletionTimestamp != nil || netPolObj.ObjectMeta.DeletionGracePeriodSeconds != nil {
+		cachedNetPolKey := util.GetNSNameWithPrefix(key)
+		err = c.cleanUpNetworkPolicy(cachedNetPolKey, SafeToCleanUpAzureNpmChain)
+		if err != nil {
+			return fmt.Errorf("Error: %v when ObjectMeta.DeletionTimestamp field is set\n", err)
+		}
+		return nil
+	}
+
+	err = c.syncAddAndUpdateNetPol(netPolObj)
+	if err != nil {
+		return fmt.Errorf("[syncNetPol] Error due to  %s\n", err.Error())
+	}
+
+	return nil
+}
+
+// initializeDefaultAzureNpmChain install default rules for kube-system and iptables
+func (c *networkPolicyController) initializeDefaultAzureNpmChain() error {
+	if c.isAzureNpmChainCreated {
+		return nil
+	}
+
+	ipsMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+	iptMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].iptMgr
+	if err := ipsMgr.CreateSet(util.KubeSystemFlag, append([]string{util.IpsetNetHashFlag})); err != nil {
+		return fmt.Errorf("[initializeDefaultAzureNpmChain] Error: failed to initialize kube-system ipset with err %s", err)
+	}
+	if err := iptMgr.InitNpmChains(); err != nil {
+		return fmt.Errorf("[initializeDefaultAzureNpmChain] Error: failed to initialize azure-npm chains with err %s", err)
+	}
+
+	c.isAzureNpmChainCreated = true
+	return nil
+}
+
+// syncAddAndUpdateNetPol handles a new network policy or an updated network policy object triggered by add and update events
+func (c *networkPolicyController) syncAddAndUpdateNetPol(netPolObj *networkingv1.NetworkPolicy) error {
+	// This timer measures execution time to run this function regardless of success or failure cases
+	timer := metrics.StartNewTimer()
+	defer timer.StopAndRecord(metrics.AddPolicyExecTime)
+
+	var err error
+	netpolKey, err := cache.MetaNamespaceKeyFunc(netPolObj)
```

Comment:
NetPoljKey right now always needs the ns- prefix right ? Why dont we create a function which gets metacachekey and adds the prefix, that way we will reduce the complexity and any changes needed will be in a single function

Reply:
if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
		utilruntime.HandleError(err)
		return
	}

	netPolCachedKey := util.GetNSNameWithPrefix(key)
this can be a single func

Reply:
Network policy controller does not need "ns-" prefix. So, we agree to not to use "ns-" prefix as key for network policy cache.

Reply:
Makes sense, we can remove the prefix in chachedNetpolKey